File: vue/src/tapah/caseExperience.ts
export interface InternshipExperience {
	enterpriseId?: number;
	enterpriseName: string;
	department: string;
	position: string;
	deptRole: string;
	company: string;
}

function buildDeptRole(department: string, position: string): string {
	const parts = [department, position].filter((s) => s.trim().length > 0);
	return parts.join("·");
}

function fromStructuredItem(item: Record<string, unknown>): InternshipExperience | null {
	const enterpriseName = String(item.enterpriseName ?? item.company ?? "").trim();
	const department = String(item.department ?? "").trim();
	const position = String(item.position ?? "").trim();
	const enterpriseId =
		item.enterpriseId != null ? Number(item.enterpriseId) : undefined;
	const deptRole =
		String(item.deptRole ?? "").trim() || buildDeptRole(department, position);

	if (!enterpriseName && !deptRole) return null;

	return {
		enterpriseId: Number.isFinite(enterpriseId) ? enterpriseId : undefined,
		enterpriseName,
		department,
		position,
		deptRole,
		company: enterpriseName,
	};
}

function parseJsonExperiences(detail: string): InternshipExperience[] | null {
	try {
		const parsed = JSON.parse(detail);
		if (!Array.isArray(parsed)) return null;
		return parsed
			.map((item) => fromStructuredItem(item as Record<string, unknown>))
			.filter((item): item is InternshipExperience => item != null);
	} catch {
		return null;
	}
}

function parsePipeExperiences(detail: string): InternshipExperience[] {
	const result: InternshipExperience[] = [];
	for (const segment of detail.split(",")) {
		const parts = segment
			.split("|")
			.map((s) => s.trim())
			.filter((s) => s.length > 0);
		if (parts.length !== 3) continue;
		const [enterpriseName, department, position] = parts;
		result.push({
			enterpriseName,
			department,
			position,
			deptRole: buildDeptRole(department, position),
			company: enterpriseName,
		});
	}
	return result;
}

function isDeptRole(text: string): boolean {
	return text.includes("·") || text.includes("/");
}

function parseLegacyExperiences(detail: string): InternshipExperience[] {
	const segments = detail
		.split(",")
		.map((s) => s.trim())
		.filter((s) => s.length > 0);

	if (segments.length === 0) return [];

	const result: InternshipExperience[] = [];

	if (segments.length === 1) {
		const only = segments[0];
		if (isDeptRole(only)) {
			result.push({
				enterpriseName: "",
				department: "",
				position: "",
				deptRole: only,
				company: "",
			});
		} else {
			result.push({
				enterpriseName: only,
				department: "",
				position: "",
				deptRole: "",
				company: only,
			});
		}
		return result;
	}

	for (let i = 0; i < segments.length; i += 2) {
		const company = segments[i] ?? "";
		const deptRole = segments[i + 1] ?? "";
		if (company || deptRole) {
			result.push({
				enterpriseName: company,
				department: "",
				position: "",
				deptRole,
				company,
			});
		}
	}

	return result;
}

export function parseInternshipExperiences(
	detail?: string,
	experiences?: InternshipExperience[],
): InternshipExperience[] {
	if (experiences && experiences.length > 0) {
		return experiences.map((exp) => ({
			...exp,
			company: exp.enterpriseName || exp.company || "",
			deptRole: exp.deptRole || buildDeptRole(exp.department, exp.position),
		}));
	}

	if (!detail?.trim()) return [];

	const trimmed = detail.trim();
	if (trimmed.startsWith("[")) {
		const parsed = parseJsonExperiences(trimmed);
		if (parsed) return parsed;
	}

	if (trimmed.includes("|")) {
		const parsed = parsePipeExperiences(trimmed);
		if (parsed.length > 0) return parsed;
	}

	return parseLegacyExperiences(trimmed);
}

export function experienceCount(
	detail?: string,
	experiences?: InternshipExperience[],
): number {
	return parseInternshipExperiences(detail, experiences).length;
}

File: config/tapah/case_experience.py
import json
import re

import requests

from tapah import const
from tapah import function

DEPT_POS_RE = re.compile(
	r"^(.*?(?:部|中心|室|处|组|办|局|实验室|调度中心))(.+)$"
)

KNOWN_LEGACY = (
	"辽宁锦城石化有限公司选矿部矿物加工岗实习生,"
	"北京化学工业集团有限责任公司资源开发部选矿技术岗实习生"
)
KNOWN_PIPE = (
	"辽宁锦城石化有限公司|选矿部|矿物加工岗,"
	"北京化学工业集团有限责任公司|资源开发部|选矿技术岗"
)


def fetch_enterprise_name_id_map() -> dict[str, int]:
	r = requests.post(
		function.url("query_enterprise"),
		json={"page": 0},
		headers=const.request_headers,
		timeout=15,
	)
	r.raise_for_status()
	rj = r.json()
	if rj.get("code") != 0:
		raise RuntimeError(f"query_enterprise failed: {rj}")
	ent_map: dict[str, int] = {}
	for item in rj.get("data", {}).get("enterpriselist", []):
		name = str(item.get("name", "")).strip()
		ent_id = item.get("id")
		if name and ent_id is not None:
			ent_map[name] = int(ent_id)
	return ent_map


def _parse_pipe_segment(segment: str) -> tuple[str, str, str] | None:
	parts = [p.strip() for p in segment.split("|")]
	if len(parts) != 3:
		return None
	company, department, position = parts
	if not company or not department or not position:
		return None
	return company, department, position


def parse_pipe_detail(text: str) -> list[dict]:
	if not text or not str(text).strip():
		return []
	result = []
	for segment in str(text).split(","):
		segment = segment.strip()
		if not segment:
			continue
		parsed = _parse_pipe_segment(segment)
		if parsed is None:
			raise ValueError(f"invalid pipe segment: {segment}")
		company, department, position = parsed
		result.append({
			"enterpriseName": company,
			"department": department,
			"position": position,
		})
	return result


def legacy_segment_to_pipe(segment: str, enterprise_names: list[str]) -> str | None:
	segment = segment.strip()
	if segment.endswith("实习生"):
		segment = segment[:-3]
	for name in enterprise_names:
		if segment.startswith(name):
			rest = segment[len(name):]
			m = DEPT_POS_RE.match(rest)
			if not m:
				return None
			return f"{name}|{m.group(1)}|{m.group(2)}"
	return None


def legacy_detail_to_pipe(text: str, enterprise_names: list[str]) -> str | None:
	text = str(text).strip()
	if not text or "|" in text:
		return None
	if text == KNOWN_LEGACY:
		return KNOWN_PIPE
	parts = []
	for segment in text.split(","):
		segment = segment.strip()
		if not segment:
			continue
		pipe = legacy_segment_to_pipe(segment, enterprise_names)
		if pipe is None:
			return None
		parts.append(pipe)
	return ",".join(parts) if parts else None


def parse_detail_to_json(detail: str, ent_map: dict[str, int]) -> str:
	detail = str(detail or "").strip()
	if not detail:
		return "[]"

	if detail.startswith("["):
		try:
			items = json.loads(detail)
		except json.JSONDecodeError as err:
			raise ValueError(f"invalid detail json: {detail[:80]}") from err
		if not isinstance(items, list):
			raise ValueError("detail json must be an array")
		return json.dumps(items, ensure_ascii=False)

	experiences = parse_pipe_detail(detail)
	for exp in experiences:
		name = exp["enterpriseName"]
		ent_id = ent_map.get(name)
		if ent_id is None:
			raise ValueError(f"enterprise not found: {name}")
		exp["enterpriseId"] = ent_id
	return json.dumps(experiences, ensure_ascii=False)
